nvbench/axes_metadata.cxx

```diff
@@ -84,26 +117,175 @@ catch (std::exception &e)
 void axes_metadata::add_float64_axis(std::string name,
                                      std::vector<nvbench::float64_t> data)
 {
-  auto axis = std::make_unique<nvbench::float64_axis>(std::move(name));
-  axis->set_inputs(std::move(data));
-  m_axes.push_back(std::move(axis));
+  this->add_axis(nvbench::float64_axis{name,data});
 }
 
 void axes_metadata::add_int64_axis(std::string name,
                                    std::vector<nvbench::int64_t> data,
                                    nvbench::int64_axis_flags flags)
 {
-  auto axis = std::make_unique<nvbench::int64_axis>(std::move(name));
-  axis->set_inputs(std::move(data), flags);
-  m_axes.push_back(std::move(axis));
+  this->add_axis(nvbench::int64_axis{name,data,flags});
 }
 
 void axes_metadata::add_string_axis(std::string name,
                                     std::vector<std::string> data)
 {
-  auto axis = std::make_unique<nvbench::string_axis>(std::move(name));
-  axis->set_inputs(std::move(data));
-  m_axes.push_back(std::move(axis));
+  this->add_axis(nvbench::string_axis{name,data});
+}
+
+void axes_metadata::add_axis(const axis_base& axis)
+{
+  m_value_space.push_back(
+    std::make_unique<linear_axis_space>(m_axes.size(),
+                                        m_axes.size() - m_type_axe_count));
+  m_axes.push_back(axis.clone());
+}
+
+namespace
+{
+std::tuple<std::vector<std::size_t>, std::vector<std::size_t>>
+get_axes_indices(std::size_t type_axe_count,
+                 const nvbench::axes_metadata::axes_type &axes,
+                 const std::vector<std::string> &names)
+{
+  std::vector<std::size_t> input_indices;
+  input_indices.reserve(names.size());
+  for (auto &n : names)
+  {
+    auto iter =
+      std::find_if(axes.cbegin(), axes.cend(), [&n](const auto &axis) {
+        return axis->get_name() == n;
+      });
+
+    // iter distance is input_indices
+    if (iter == axes.cend())
+    {
+      NVBENCH_THROW(std::runtime_error,
+                    "Unable to find the axes named ({}).",
+                    n);
+    }
+    auto index = std::distance(axes.cbegin(), iter);
+    input_indices.push_back(index);
+  }
+
+  std::vector<std::size_t> output_indices = input_indices;
+  for (auto &out : output_indices)
+  {
+    out -= type_axe_count;
+  }
+  return std::tie(input_indices, output_indices);
```

Comment:
`tie` isn't what you want here. Since the return type is explicit, we can just implicitly construct it. We can also avoid copying the vectors. 
```suggestion
  return {std::move(input_indices), std::move(output_indices)};
```

Reply:
Awesome, thanks.